29.md

```diff
@@ -6,153 +6,270 @@ Relay-based Groups
+Access control queries and requests are implemented using new message types, rather than events, in order to prevent broadcasting of sensitive data by relays that don't implement NIP 29. Any relay, regardless of whether they support NIP 29, MAY implement these verbs using an empty string as the group id.
+
+Users MAY query for their current permissions using the `ACCESS` verb with an optional `group id`.
+Relays MAY respond with zero or more messages using the same `ACCESS` verb.
+
+- The first argument MUST be a group id.
+- The second argument MUST be a comma-separated list of permissions, either `read` or `write`.
+- The third argument MAY be a human-readable message to be displayed to the user.
 
 ```json
-["ACCESS"]
+-> ["ACCESS"]
+<- ["ACCESS", "", "read,write"]
+<- ["ACCESS", "_", "read,write"]
+<- ["ACCESS", "<group1-id>", "read"]
+<- ["ACCESS", "<group2-id>", "", "We don't like your kind around here"]
 ```
 
-In either case, a relay MUST respond with an `ACCESS` message with a list of user permissions, in addition to the standard `OK` message. Some examples:
+Users MAY request access using a `JOIN` message sent ONLY to the relay in question. Relay access requests MUST have a `group id` and MAY have a `claim` containing an arbitrary string (such as an invite code, zap receipt, cashu token, etc). A message to the relay admin MAY be included.
 
 ```json
-["ACCESS"]
-["ACCESS", "read"]
-["ACCESS", "read", "write"]
+["JOIN", "<group-id>", "<claim>", "<message>"]
 ```
 
-Valid permissions are `read` and `write`.
+Users MAY also request that their access be revoked using a `LEAVE` messsage:
 
-## Moderation
+```json
+["LEAVE"]
+```
 
-Relay-based groups have a single admin, identified by the `pubkey` listed in the relay's [NIP 11](11.md) document.
+Relays MAY update internal state and member lists, or ignore these messages.
 
-Any relay member MAY publish a `kind 1984` report to the relay (including a `-` tag if desired/supported). These reports MAY be used by the relay admin to delete events or ban users, or by clients to implement any moderation algorithm desired. Admins or moderators may also choose to escalate reports without banning content by publishing their own `kind 1984` event with the same tags.
+# Membership
 
-## Membership
+## Group lists
 
-Users MAY track their own group memberships using a [NIP 51](51.md) kind `1xxxx` event. Tags MAY be either public or encrypted with [NIP 44](44.md), depending on user/client preference.
+Users MAY track their own group memberships using a [NIP 51](51.md) kind `10009` event. Tags MAY be either public or encrypted with [NIP 44](44.md), depending on user/client preference.
 
-Membership lists can contain two types of tags: relay memberships (indicated by `r`), and room memberships (indicated by `h`, and including the relay url):
+Group IDs are identified by `group`, with each value being the relay url and the group id joined by a `'`. Group id MAY be omitted if bare relay membership is desired.
```

Comment:
why not just another argument?

`[ "group", "group-id", "relay" ]`

Reply:
See the note below, that would be my preference.